h5py/_hl/vds.py

```diff
@@ -102,6 +103,20 @@ def shape(self):
     def __getitem__(self, key):
         tmp = copy(self)
         tmp.sel = select(self.shape, key, dataset=None)
+        try:
+            key = tuple(key)
+        except TypeError:
+            key = (key,)
+        for i, sl in enumerate(key):
+            if isinstance(sl, slice):
+                if sl.stop == h5s.UNLIMITED:
+                    space = tmp.sel.id
+                    rank = space.get_simple_extent_ndims()
+                    start, stride, count, block = space.get_regular_hyperslab()
+                    counts = list(count)
+                    counts[i] = h5s.UNLIMITED
```

Comment:
Unfortunately, just iterating through and modifying the `i-th` index won't quite work, because you can use ellipsis, like `obj[..., :UNLIMITED]` to slice the last dimension.

`...` is equivalent to filling any missing dimensions (`rank - (len(args) - 1)`) with `slice(None, None)`, which is equivalent to `[:]`. We should be able to rely on `select()` throwing an error if it's used incorrectly (e.g. multiple ellipsis).

Reply:
Yeah, I was doing the selection changes after using `select()` to catch those kinds of things. I don't think `select()` should be changed at all since the `UNLIMITED` hyperslab selection was added specifically for VDS, and I'm not sure if it actually works with other kinds of selections.

Reply:
Thanks, that makes sense. The main use of selection is for reading & writing data, so I think you're right that UNLIMITED only makes sense for virtual dataset mappings, and thus should be handled here. But hopefully we can skip some of the sanity checks around ellipsis, because `select()` already does them.